tsfresh/feature_extraction/extraction.py

```diff
@@ -103,9 +114,103 @@ def extract_features(timeseries_container, feature_extraction_settings=None,
+    :type timeseries_container: pandas.DataFrame or dict
+
+    :param column_id: The name of the id column to group by.
+    :type column_id: str
+    :param column_value: The name for the column keeping the value itself.
+    :type column_value: str
+
+    :param settings: settings object that controls which features are calculated
+    :type settings: tsfresh.feature_extraction.settings.FeatureExtractionSettings
+
+    :return: The (maybe imputed) DataFrame containing extracted features.
+    :rtype: pandas.DataFrame
+    """
+    all_possible_unique_id_values = set(id_value for kind, df in kind_to_df_map.items()
+                                        for id_value in df[column_id])
+    df_with_ids = pd.DataFrame(index=all_possible_unique_id_values)
+
+    partial_extract_features_for_one_time_series = partial(_extract_features_for_one_time_series,
+                                                           column_id=column_id,
+                                                           column_value=column_value,
+                                                           settings=settings)
+    pool = Pool(settings.n_processes)
+
+    extracted_features = pool.map(partial_extract_features_for_one_time_series, kind_to_df_map.items(),
+                                  chunksize=settings.chunksize)
+
     pool.close()
+
+    # Concatenate all partial results
+    result = pd.concat([df_with_ids] + extracted_features, axis=1, join='outer', join_axes=[df_with_ids.index]) \
+        .astype(np.float64)
+
     pool.join()
+    return result
+
+
+def _extract_features_parallel_per_sample(kind_to_df_map, settings, column_id, column_value):
+    """
+    Parallelize the feature extraction per kind and per sample.
+
+    :param kind_to_df_map: The time series to compute the features for in our internal format
+    :type kind_to_df_map: dict of pandas.DataFrame
+
+    :param column_id: The name of the id column to group by.
+    :type column_id: str
+    :param column_value: The name for the column keeping the value itself.
+    :type column_value: str
+
+    :param settings: settings object that controls which features are calculated
+    :type settings: tsfresh.feature_extraction.settings.FeatureExtractionSettings
 
+    :return: The (maybe imputed) DataFrame containing extracted features.
+    :rtype: pandas.DataFrame
+    """
+    all_possible_unique_id_values = set(id_value for kind, df in kind_to_df_map.items()
+                                        for id_value in df[column_id])
+    df_with_ids = pd.DataFrame(index=all_possible_unique_id_values)
+
+    partial_extract_features_for_one_time_series = partial(_extract_features_for_one_time_series,
+                                                           column_id=column_id,
+                                                           column_value=column_value,
+                                                           settings=settings)
+    pool = Pool(settings.n_processes)
```

Comment:
If we can get rid of the ``df_with_ids`` only the creation of the partial and the setup of the Pool are common.
I would definitely leave the pool inside the functions as it enables us to later switch to e.g. joblib for per sample and keep the Pool for per kind. Also I wouldn't seperate the setup and closing of the pool into seperate functions.
For the partial you might be right, but it is one line and we might want to set it up differently in the future.